src/DiagramGenerator/Image/Storage.php

```diff
@@ -0,0 +1,189 @@
+<?php
+
+namespace DiagramGenerator\Image;
+
+use DiagramGenerator\Config;
+use DiagramGenerator\Config\Texture;
+use DiagramGenerator\Fen;
+use DiagramGenerator\Fen\Piece;
+use Intervention\Image\Exception\NotReadableException;
+use Intervention\Image\Image;
+use Intervention\Image\ImageManagerStatic;
+use RuntimeException;
+
+class Storage
+{
+    protected $pieces = [];
+
+    /** @var string */
+    protected $cacheDirectory;
+
+    /** @var string */
+    protected $pieceThemeUrl;
+
+    /** @var string */
+    protected $boardTextureUrl;
+
+    /**
+     * @param string $cacheDirectory
+     * @param string $pieceThemeUrl
+     * @param string $boardTextureUrl
+     */
+    public function __construct($cacheDirectory, $pieceThemeUrl, $boardTextureUrl)
+    {
+        $this->cacheDirectory = sprintf('%s/diagram_generator', $cacheDirectory);
+        $this->pieceThemeUrl = $pieceThemeUrl;
+        $this->boardTextureUrl = $boardTextureUrl;
+    }
+
+    /**
+     * @param Piece  $piece
+     * @param Config $config
+     *
+     * @return Image
+     */
+    public function getPieceImage(Piece $piece, Config $config)
+    {
+        $cacheKey = implode('.', [$piece->getColor(), $piece->getKey(), $piece->getColumn(), $piece->getRow()]);
+
+        if (!isset($this->pieces[$cacheKey])) {
+            $this->pieces[$cacheKey] = $this->fetchRemotePieceImage($piece, $config);
+        }
+
+        return $this->pieces[$cacheKey];
+    }
+
+    /**
+     * @param Config $config
+     *
+     * @return Image
+     */
+    public function getBackgroundTextureImage(Config $config)
+    {
+        $boardCachedPath = $this->getCachedTextureFilePath($config);
+
+        try {
+            return ImageManagerStatic::make($boardCachedPath);
+        } catch (NotReadableException $exception) {
```

Comment:
Yeah. Vendor basically calls `getimagesize()`, and if it fails, throws this exception. So, it covers all your cases.